reco_utils/dataset/pandas_df_utils.py

```diff
@@ -74,3 +75,75 @@ def filter_by(df, filter_by_df, filter_by_cols):
             filter_by_df.set_index(filter_by_cols).index
         )
     ]
+
+
+def df_to_libffm(df, col_rating=DEFAULT_RATING_COL, filepath=None):
+    """Converts an input Dataframe (df) to an text file in libffm format.
+
+    Note:
+        The input dataframe is expected to represent the feature data in the following schema
+        |field-1|field-2|...|field-n|rating|
+        |feature-1-1|feature-2-1|...|feature-n-1|1|
+        |feature-1-2|feature-2-2|...|feature-n-2|0|
+        ...
+        |feature-1-i|feature-2-j|...|feature-n-k|0|
+        Where
+        1. each "field-*" occupies one column in the data, and
+        2. "feature-*-*" can be either a string or a numerical value.
+
+        The above data will be converted to the libffm format by following the convention as explained in
+        https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
+
+    Args:
+        df (pd.DataFrame): input Pandas dataframe.
+        col_rating (str): rating of the data.
+        filepath (str): path to save the converted data.
+
+    Return:
+        df_libffm (pd.DataFrame): a Pandas DataFrame in libffm format.
+    """
+    df_new = df.copy()
+
+    # Encode field (rating column is not included)
+    field_names = list(df_new.drop(col_rating, axis=1).columns)
+    field_dict = {k: v+1 for v, k in enumerate(field_names)}
+
+    # Encode field-feature
+    field_feature_list = []
+    for col in field_names:
+        if df_new[col].dtype == int:
+            df_new[col] = df_new[col].astype(str)
```

Comment:
why is this converted to string? don't you want to retain handling this value numerically in _convert?

Reply:
+1
Shouldn't floating types and integers treated in the same way?

Reply:
Some times the column values can be represented in `int` type while they are actually ordinal variables or something similar.

Reply:
I think it is impossible, in general, to write a "universal" data preprocessor or converter. There is always an ad hoc element in these things. 
So, I think it is better to narrow down what is the context in which you need / will use this util:
- what parts of the code use it?
- document exactly what it does i.e. how does it extract features and fields from the DF, how it treats different data types etc. so that the user does not need to look into the code to figure what is going on.
- rename it to something less generic; in many cases the util will not do what the data scientist wants (and that is ok because there is no easy way to write a universal util).

Reply:
> So, I think it is better to narrow down what is the context in which you need / will use this util

Completely agree. This is what I amended in the most recent update - to allow only `str` and `numeric` types whilst the latter is regarded as numeric variable regardless it is int, float, or anything else. 

> what parts of the code use it? e.g. can it be used for DeepRec?

I have tested with the xDeepFM and it works well. The use case is to convert a data that is usually represented in dataframe with feature columns to be a format that is required by the APIs in DeepRec algorithms (or any package that reads a libffm format file).

> document exactly what it does i.e. how does it extract features and fields from the DF, how it treats different data types etc. so that the user does not need to look into the code to figure what is going on.

This was texted in the docstring. However, I am not sure if the conversion logic is needed as the reference is given where it is explained very clearly.

> rename it to something less generic; in many cases the util will not do what the data scientist wants (and that is ok because there is no easy way to write a universal util).

Make sense. Any suggestion here?

Reply:
Maybe the best way would be to demonstrate by example. I.e. show an input DF, the resulting output DF and what is written to file.

Reply:
Yea this is good to have. Added.